src/NzbDrone.Core/Download/History/DownloadHistoryService.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using NzbDrone.Common.Extensions;
+using NzbDrone.Core.History;
+using NzbDrone.Core.MediaFiles.Events;
+using NzbDrone.Core.Messaging.Events;
+using NzbDrone.Core.Tv.Events;
+
+namespace NzbDrone.Core.Download.History
+{
+    public interface IDownloadHistoryService
+    {
+        bool DownloadAlreadyImported(string downloadId);
+        DownloadHistory GetLatestDownloadHistoryItem(string downloadId);
+    }
+
+    public class DownloadHistoryService : IDownloadHistoryService,
+                                          IHandle<EpisodeGrabbedEvent>,
+                                          IHandle<EpisodeImportedEvent>,
+                                          IHandle<DownloadCompletedEvent>,
+                                          IHandle<DownloadFailedEvent>,
+                                          IHandle<DownloadIgnoredEvent>,
+                                          IHandle<SeriesDeletedEvent>
+
+    {
+        private readonly IDownloadHistoryRepository _repository;
+        private readonly IHistoryService _historyService;
+
+        public DownloadHistoryService(IDownloadHistoryRepository repository, IHistoryService historyService)
+        {
+            _repository = repository;
+            _historyService = historyService;
+        }
+
+        public bool DownloadAlreadyImported(string downloadId)
+        {
+            var events = _repository.FindByDownloadId(downloadId);
+
+            // Events are ordered by date descending, if a grabbed event comes before an imported event then it was never imported
```

Comment:
Nice, but that could cause issues with partial imports that never finished, and was later regrabbed. Those FileImported events would still be there and potentially picked up. (At least that's what I assume FileImported would eventually be used for)
We could deal with that by comparing the Data of the FileImported events to the grab event and ignore them.